utils/prometheus/src/sourced.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metrics that are collected from existing sources.
+
+use prometheus::core::{Collector, Desc, Describer, Number, Opts};
+use prometheus::proto;
+use std::{cmp::Ordering, marker::PhantomData};
+
+/// A counter whose values are obtained from an existing source.
+pub type SourcedCounter<S> = SourcedMetric<Counter, S>;
+
+/// A gauge whose values are obtained from an existing source.
+pub type SourcedGauge<S> = SourcedMetric<Gauge, S>;
+
+/// The type of a sourced counter.
+#[derive(Copy, Clone)]
+pub enum Counter {}
+
+/// The type of a sourced gauge.
+#[derive(Copy, Clone)]
+pub enum Gauge {}
+
+/// A metric whose values are obtained from an existing source,
+/// instead of being independently recorded.
+#[derive(Debug, Clone)]
+pub struct SourcedMetric<T, S> {
+	source: S,
+	desc: Desc,
+	_type: PhantomData<T>,
+}
+
+/// A source of values for a [`SourcedMetric`].
+pub trait MetricSource: Sync + Send + Clone {
+	/// The type of the collected values.
+	type N: Number;
+	/// Collects the current values of the metrics from the source.
+	fn collect(&self, set: impl FnMut(&[&str], Self::N));
+}
+
+impl<T: SourcedType, S: MetricSource> SourcedMetric<T, S> {
+	/// Creates a new metric that obtains its values from the given source.
+	pub fn new(opts: &Opts, source: S) -> prometheus::Result<Self> {
+		let desc = opts.describe()?;
+		Ok(Self { source, desc, _type: PhantomData })
+	}
+}
+
+impl<T: SourcedType, S: MetricSource> Collector for SourcedMetric<T, S> {
+	fn desc(&self) -> Vec<&Desc> {
+		vec![&self.desc]
+	}
+
+	fn collect(&self) -> Vec<proto::MetricFamily> {
+		let mut counters = Vec::new();
+
+		self.source.collect(|label_values, value| {
+			let mut m = proto::Metric::default();
+
+			match T::proto() {
+				proto::MetricType::COUNTER => {
+					let mut c = proto::Counter::default();
+					c.set_value(value.into_f64());
+					m.set_counter(c);
+				}
+				proto::MetricType::GAUGE => {
+					let mut g = proto::Gauge::default();
+					g.set_value(value.into_f64());
+					m.set_gauge(g);
+				}
+				t => {
+					log::error!("Unsupported sourced metric type: {:?}", t);
+				}
+			}
+
+			match self.desc.variable_labels.len().cmp(&label_values.len()) {
```

Comment:
All rust-prometheus methods panic when providing unequal label keys and values. While not the greates UX, I would do the same (at least in debug mode) to stay consistent.

E.g. https://docs.rs/prometheus/0.7.0/prometheus/core/struct.MetricVec.html#method.with_label_values.

Reply:
Done in https://github.com/paritytech/substrate/pull/6895/commits/45b734cce4708839bf290c0d872b4cb1d690fd52.